api/python/quilt3/data_transfer.py

```diff
@@ -298,10 +298,15 @@ def upload_part(i, start, end):
                     Bucket=dest_bucket,
                     Key=dest_key,
                     UploadId=upload_id,
-                    PartNumber=part_id
+                    PartNumber=part_id,
+                    ChecksumAlgorithm='SHA256',
                 )
             with lock:
-                parts[i] = {"PartNumber": part_id, "ETag": part["ETag"]}
+                parts[i] = dict(
```

Comment:
I was expecting at least one case in which we need to ListParts or know the part size from S3? E.g. what about `quilt3 verify` or any case where we only have the checksum and the file and need to verify it.

Reply:
We _could_ drop support for quilt3 verify if this is a pain, but it's nice. And actually checksums should make it *much* faster.

Reply:
`ListParts` requires an upload ID - and I'm not sure it'll even work for completed uploads (though I didn't check).

In any case, I got `verify` working - but, it relies on the assumption that the files were uploaded using the default values for `multipart_threshold` and `multipart_chunksize` (both 8MB). We're always using the defaults - but, if somehow the user uploads the files using other tools _and_ changes the defaults _and_ enables the sha256 checksum, then they won't match. But that seems like a pretty extreme edge case.

Reply:
A list of parts along with their checksums cant be obtained like this:
```
aws s3api get-object-attributes --bucket quilt-sergey-test-promote --key 32MB-file --object-attributes ObjectParts,Checksum
```
```
{
    "LastModified": "Thu, 07 Apr 2022 14:50:33 GMT",
    "VersionId": "fR2lZdKdWx5YF4C6hbpN6Hx7MzB6svcv",
    "Checksum": {
        "ChecksumSHA256": "fMJJ08ZgZU9Z8aMET6oXADfcmluSqcXXtR07Xy3DCoQ="
    },
    "ObjectParts": {
        "TotalPartsCount": 2,
        "PartNumberMarker": 0,
        "NextPartNumberMarker": 2,
        "MaxParts": 1000,
        "IsTruncated": false,
        "Parts": [
            {
                "PartNumber": 1,
                "Size": 17179870,
                "ChecksumSHA256": "DNRfQslATbmfAaQNE15vP9kXaATe/NoGJ5Di2pkDIOc="
            },
            {
                "PartNumber": 2,
                "Size": 16374562,
                "ChecksumSHA256": "ZekGMfeF9+gphL8RrAL5ppwFWfRdNLO73mx6i1qph2M="
            }
        ]
    }
}
```

Reply:
(Just going through existing comments - want to either fix it or punt it.)

`get-object-attributes` does appear to work - but only for objects in S3, obviously. I think the normal use-case for `verify` is to check if a local directory has correct checksums - but, once files are downloaded locally, we lose all the object parts info. (We _could_ special case it for a remote package, with S3 physical keys, and use those to get object parts, but then use local files to calculate the hash... but this seems a bit crazy.)

Reply:
> In any case, I got verify working - but, it relies on the assumption that the files were uploaded using the default values for multipart_threshold and multipart_chunksize (both 8MB). We're always using the defaults - but, if somehow the user uploads the files using other tools and changes the defaults and enables the sha256 checksum, then they won't match. But that seems like a pretty extreme edge case.

I haven't re-checked PR yet, so I'm not sure how that comment relates to the existing code, but if we get existing checksum from S3 object we can't *assume* it's compliant, we **MUST** check object has compliant part sizes.

Reply:
Right now, the code doesn't rely on existing checksums at all, compliant or not - and only uses them when uploading to S3. When an existing S3 file gets added to a package, it's downloaded and hashed locally.

Reply:
Okay, that's fine. I will document that this differs from the Catalog that way.